src/jit/optimizer.cpp

```diff
@@ -6203,6 +6203,13 @@ bool Compiler::optHoistLoopExprsForTree(
         {
             treeIsHoistable = false;
         }
+
+        if (tree->OperGet() == GT_CLS_VAR)
```

Comment:
I'd suggest moving this up a few lines and sinking the call to `optVNIsLoopInvariant` into an `else` path, to avoid that call when we're just going to discard the result.  Then you should be able to simply remove the partial workaround above, since the check on 6204 will set `treeIsHoistable` to false if you're setting `treeIsInvariant` to false before it.